### DISCUSSION TITLE
superRefine don't work

### DISCUSSION THREAD

**Discussion by f1eeman** [colinhacks/zod#3296] [Q&A] [answered]
`console.log('superRefine',values) `  doesn't even occur

```
export const getBaseManagementGateFormSchema = () =>
  z
    .object({
     ....
    })
    .superRefine((values, ctx) => {
      console.log('superRefine', values)
       // never gets called
      }
    })
```

**Comment by JacobWeisenburger** [comment, marked as answer]
Here is a much simpler example:

```ts
const schema = z.object( {
    foo: z.string()
} ).superRefine( ( values, ctx ) => {
    console.log( 'superRefine', values )
    // never gets called
} )

const result = schema.safeParse( {} )
!result.success && console.log( result.error.issues )
// [
//     {
//         code: "invalid_type",
//         expected: "string",
//         received: "undefined",
//         path: [ "foo" ],
//         message: "Required",
//     }
// ]
```

Not sure if this is a bug or intended behavior

**Comment by JacobWeisenburger** [comment]
Why did you close this?

**Comment by f1eeman** [comment]
> Here is a much simpler example:
> 
> ```ts
> const schema = z.object( {
>     foo: z.string()
> } ).superRefine( ( values, ctx ) => {
>     console.log( 'superRefine', values )
>     // never gets called
> } )
> 
> const result = schema.safeParse( {} )
> !result.success && console.log( result.error.issues )
> // [
> //     {
> //         code: "invalid_type",
> //         expected: "string",
> //         received: "undefined",
> //         path: [ "foo" ],
> //         message: "Required",
> //     }
> // ]
> ```
> 
> Not sure if this is a bug or intended behavior

Try this:
```
const schema = z.object( {
    foo: z.string().optional()
} ).superRefine( ( values, ctx ) => {
    console.log( 'superRefine', values )
    // never gets called
} )
```

**Comment by f1eeman** [comment]
> `console.log('superRefine',values) ` doesn't even occur
> 
> ```
> export const getBaseManagementGateFormSchema = () =>
>   z
>     .object({
>      ....
>     })
>     .superRefine((values, ctx) => {
>       console.log('superRefine', values)
>        // never gets called
>       }
>     })
> ```

work for me:

```
export const getBaseManagementGateFormSchema = () => {
  const schema = z
    .object({
     ...
    })
    
    return schema.superRefine((values, ctx) => {
      ...
    })
}
```

**Reply by MichaelDimmitt** [reply]
```typescript
const schema = z
    .object({
      foo: z.string(),
    })
    .superRefine((values, ctx) => {
      console.log('superRefine', values);
      // never gets called
    });

const result = schema.safeParse({ foo: 'hi' });
```
Just an outside observation but that super refine was never being reached because the it was failing due to z.string being a required field. That happens before the super refine is hits so it looks like it short circuits.

The code snippet I posted sends an object with foo having a string value and the super refine does get hit.
@JacobWeisenburger @f1eeman
